internal/saucecloud/cloud.go

```diff
@@ -934,3 +967,128 @@ func (r *CloudRunner) getHistory(launchOrder config.LaunchOrder) (insights.JobHi
 	}
 	return r.InsightsService.GetHistory(context.Background(), user, launchOrder)
 }
+
+func (r *CloudRunner) reportSuiteToInsights(res result) {
+	// Skip reporting if job is not completed
+	if !job.Done(res.job.Status) || res.skipped || res.job.ID == "" {
+		return
+	}
+
+	assets, err := r.JobService.GetJobAssetFileNames(context.Background(), res.job.ID, res.job.IsRDC)
+	if err != nil {
+		log.Warn().Err(err).Str("action", "loadAssets").Msg(msg.InsightsReportError)
+		return
+	}
+
+	var testRuns []insights.TestRun
+
+	if arrayContains(assets, saucereport.SauceReportFileName) {
+		report, err := r.loadSauceTestReport(res.job.ID, res.job.IsRDC)
+		if err != nil {
+			log.Warn().Err(err).Str("action", "parsingJSON").Msg(msg.InsightsReportError)
+			return
+		}
+		testRuns = insights.FromSauceReport(report)
+	} else if arrayContains(assets, junit.JunitFileName) {
+		report, err := r.loadJUnitReport(res.job.ID, res.job.IsRDC)
+		if err != nil {
+			log.Warn().Err(err).Str("action", "parsingXML").Msg(msg.InsightsReportError)
+			return
+		}
+		testRuns = insights.FromJUnit(report)
+	}
+	enrichInsightTestRun(testRuns, res.job.ID, res.name, res.details, res.job.IsRDC)
+	if len(testRuns) > 0 {
+		if err := r.InsightsService.PostTestRun(context.Background(), testRuns); err != nil {
+			log.Warn().Err(err).Str("action", "posting").Msg(msg.InsightsReportError)
+		}
+	}
+}
+
+func (r *CloudRunner) loadSauceTestReport(jobID string, isRDC bool) (saucereport.SauceReport, error) {
+	fileContent, err := r.JobService.GetJobAssetFileContent(context.Background(), jobID, saucereport.SauceReportFileName, isRDC)
+	if err != nil {
+		log.Warn().Err(err).Str("action", "loading-json-report").Msg(msg.InsightsReportError)
+		return saucereport.SauceReport{}, err
+	}
+	return saucereport.Parse(fileContent)
+}
+
+func (r *CloudRunner) loadJUnitReport(jobID string, isRDC bool) (junit.TestSuites, error) {
+	fileContent, err := r.JobService.GetJobAssetFileContent(context.Background(), jobID, junit.JunitFileName, isRDC)
+	if err != nil {
+		log.Warn().Err(err).Str("action", "loading-xml-report").Msg(msg.InsightsReportError)
+		return junit.TestSuites{}, err
+	}
+	return junit.Parse(fileContent)
+}
+
+func arrayContains(list []string, want string) bool {
+	for _, item := range list {
+		if item == want {
+			return true
+		}
+	}
+	return false
+}
+
+func enrichInsightTestRun(runs []insights.TestRun, jobID string, jobName string, details insightDetails, isRDC bool) {
```

Comment:
Imho it'd be better to move this logic to the insights package. That way, this business logic is not spread out across multiple packages (easier to maintain).

Perhaps adjusting the signature of `insights.FromSauceReport(report)` and `insights.FromJUnit(report)` so that you get a fully filled out TestRun report struct in one call is beneficial as well.

Reply:
Updated !